TEST: Add examples for different endpoints through langchain

govarun opened the pull request:
Have added examples for hitting different endpoints, thus showing the versatile use cases of the plugin through langchain.

ngundotra commented:
Hey Varun, thanks for this PR. 

There are a few errors with this.
1. `How many lamports does .... have?`
I noticed that example 1 gives 0.0425 lamports for the address, but that's wrong. The actual number is 0.0425 * 1e9 lamports. The response from the server even explicitly says "sol". What are some ways you think we can fix this?

2. `Balance in account ... have?`
This question doesn't make that much sense - are you expecting a different answer from example 1?

This also fails to parse actions after an initial attempt to query the account info.
```
> Entering new AgentExecutor chain...
We need to use the Solana API to get the balance of the account.
Action: solana
Action Input: {"path": "/accounts/8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625", "method": "get"}
Observation: # HUGE NEXT JS error
```
Do you have some ideas on how to get around this, or at least make this more consistent?

govarun commented:
Thanks @ngundotra , for the review and the comments. Addressed issues: 

1. Have rephrased the question now, so it gives the balance in SOL and also says that the balance is in SOL. 
2. The second question had a typo and was also redundant given the first one. I have removed it now. 
Let me know how does this one look

ngundotra commented on langchain_examples/python/main.py:
```diff
@@ -1,30 +1,48 @@
 from langchain.chat_models import ChatOpenAI
```
Let's rename this file to `basic_examples.py`

ngundotra commented on langchain_examples/python/plugintool.py:
```diff
@@ -0,0 +1,173 @@
+                    )
+
+        base_str = ""
+        for op in operations:
+            op_str = f"{op.path}: ({op.method}) {op.description}\n"
+            # TODO: add support and tests for query_params
```
Can you add tests for this before we merge this in?

I want to make sure we narrow down any possible bugs in this class before moving on

ngundotra commented on langchain_examples/python/main.py:
```diff
@@ -1,30 +1,48 @@
-    agent_chain.run(
-        "How many lamports does 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625 have?")
+    # Case 1 (getBalance):
+    agent_chain.run("What is the balance in account 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625?")
```
Reword to `What is the balance of 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625?`

ngundotra commented on langchain_examples/python/main.py:
```diff
@@ -1,30 +1,48 @@
+    # Expected response: <Numerical value>
+    
+    # Case 2 (getAccountInfo): 
+    agent_chain.run("Could you some info for the account 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625?")
```
`Could you provide some info for account 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625?`

ngundotra commented on langchain_examples/python/main.py:
```diff
@@ -1,30 +1,48 @@
+    # Expected response: The account 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625 has a balance of <Numerical value> lamports and is owned by the address <owner>.
+
+    # Case 3 (getTokenAccounts): 
+    agent_chain.run("Could you tell me all the assets owned by owner 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625 ?")
```
This is misleading - I actually see it hit `getAssetsForOwner`, not `getTokenAccounts`. 
In actuality, we probably want both endpoints to get hit & aggregated.

govarun replied:
The chain would probably stop just at one of them, if it gets the info. I'll drop a TODO as a future exploration to see how we can hit two endpoints and accumulate the info

ngundotra commented on langchain_examples/python/main.py:
```diff
@@ -1,30 +1,48 @@
+    
+
+    # Case 4 (getTokenAccounts): 
+    agent_chain.run("Could you tell me the last transaction i made on Solana with address 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625?")
```
`Can you tell me the last transaction made by 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625`?

ngundotra commented on langchain_examples/python/main.py:
```diff
@@ -1,30 +1,48 @@
+    # Expected response: The assets owned by the owner with address 8fbqVvpK3Dj7fdP2c8JJhtD7Zy3n9qtwAeGfbkgPu625 are [{"mint":<Address>,"amount":<Numerical Value>}].
+    
+
+    # Case 4 (getTokenAccounts): 
```
This is the wrong comment... it actually should be hitting `getSignaturesForAddress`

ngundotra reviewed (changes requested):
1. Small nits & grammatical issues
2. Add test for input -> post requests made by the PluginTool
3. Add a test for each of the questions in `basic_eamples.py` - I want to know what the baseline is for answering these questions. Off the cuff guess is that the agent should be able to answer these questions (or at least hit the right endpoint) 3/5 times - so run each in a for loop & try catch & benchmark this in a separate folder of tests please.

govarun commented:
Examples have been added to the demo.
